graph/patterns/alternate-key.md

```diff
@@ -0,0 +1,135 @@
+# Alternate Key Pattern
+
+Microsoft Graph API Design Pattern
+
+_The Alternate Key Pattern provides the ability to query for a single, specific resource identifiable through an alternative set of properties that is not its primary key_
+
+## Problem
+
+---
+
+The resources exposed in Graph are identified through a Primary Key - which guarantees uniqueness inside the same resource type. Often though, that same resource can also be uniquely identified by an alternative, more convenient property (or set of properties) that provides a better developer experience.
+
+Take a look at the `user` resource: while the `id` remains a perfectly valid way to get the resource details, the `mail` address is also an unique property that could be used to identify it.
+
+While it is still possible to use the OData filter, such as
+
+`https://graph.microsoft.com/v1.0/users?$filter=mail eq 'bob@contoso.com'`, the returned result is wrapped in an array that needs to be unpacked.
+
+## Solution
+
+---
+
+OData offers resource addressing via an alternate key using the same parentheses-style convention as for the canonical key, with one difference: single-part alternate keys MUST specify the key property name to unambiguously determine the alternate key. (Note: this is a hypothetical sample)
+
+https://graph.microsoft.com/v1.0/users(0) - Retrieves the employee with ID = 0
+https://graph.microsoft.com/v1.0/users(email='bob@contoso.com') Retrieves the employee with the email matching `bob@contoso.com`
+
+## When to Use this Pattern
+
+---
+
+This pattern works and makes sense when the alternate key is good enough to identify a single resource and provides an useful alternative to the client; while it does not work if the resultset has more than one element.
+
+In such case, the workload has two choices:
+
+1. If the workload defines a list of alternate keys for a resource and the requested key is not there, the workload SHOULD return `422`
+2. If the workload does not have a list of alternate keys, it can still query the data source, but if the result yields more than a record, it SHOULD return `422` and encourage the user to use `$filter` rather than returning the first result of the query
```

Comment:
This seems problematic here in the scenario where an entity is adding to the collection which makes the alternative key non-unique e.g.
1) Caller makes a request for https://graph.microsoft.com/v1.0/users(email='bob@contoso.com') and gets a 200
2) Another user is added who also has the email of bob@contoso
3) Caller makes a request for https://graph.microsoft.com/v1.0/users(email='bob@contoso.com') and now gets a 422

In my opinion a workload should only support an alternative key if that key is guaranteed to be unique within the dataset, which would imply that we should only support explicitly declared alternative keys

Reply:
FYI the [Odata spec](https://github.com/OData/vocabularies/blob/master/OData.Community.Keys.V1.md#223key-value-uniqueness) says

> Alternate Keys are intended to allow callers to refer to a specific entity instance. Therefore, services must allow 1 and only 1 record in an entity collection to match any set of values for the set of properties defined for each key.

Reply:
You are correct. I was not aware that alternative keys must be declared in advance in the CSDL file